### PR TITLE
Added normalization, reduction and atomic premitieves

### PR DESCRIPTION
**Pull request by AvKhokhlov**

### CODE REVIEW

File: CLW/CL/CLW.cl [hunk 1]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
         }
     }
 }
+
+// --------------------- ATOMIC OPERTIONS ------------------------
+
+#define DEFINE_ATOMIC(operation)\
+    inline void atomic_##operation##_float(volatile __global float* addr, float value)\
+    {\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: Fix formattting, { from new line

File: CLW/CL/CLW.cl [hunk 2]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+#define DEFINE_ATOMIC_FLOAT3(operation)\
+    inline void atomic_##operation##_float3(volatile __global float3* addr, float3 value)\
+    {\
+        volatile __global float* p = (volatile __global float*)addr;\
+        atomic_##operation##_float(p, value.x);\
+        atomic_##operation##_float(p + 1, value.y);\
+        atomic_##operation##_float(p + 2, value.z);\
+    }
+
+inline void atomic_max_int(volatile __global int* addr, int value)
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: `inline -> __attribute__((always_inline))`

File: CLW/CL/CLW.cl [hunk 3]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+
+inline void atomic_min_int(volatile __global int* addr, int value)
+{
+    atomic_min(addr, value);
+}
+
+// --------------------- HELPERS ------------------------
+
+#define DEFINE_ASSIGN_OPERATOR(type)\
+    inline void assign_##type(__local type* addr, type value)\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: why do we need this function? would *addr = value work?

File: CLW/CL/CLW.cl [hunk 4]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+{
+    (*addr).xyz = value.xyz;
+}
+
+inline int divide_int(int dividend, int divider)
+{
+    return dividend / (divider != 0 ? divider : 1);
+}
+
+inline float divide_float(float dividend, float divider)
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: This divide operations look weird, why IEEE floating point division is not enough?

File: CLW/CL/CLW.cl [hunk 5]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+    return (float3)(divide_float(dividend.x, divider.x),
+                    divide_float(dividend.y, divider.y),
+                    divide_float(dividend.z, divider.z));
+} 
+
+// --------------------- REDUCTION ------------------------
+
+#define DEFINE_REDUCTION(bin_op, type)\
+__kernel void reduction_##bin_op##_##type(__global type* buffer,\
+                                          int buf_count,\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: bad name, use simply count or num_elems

File: CLW/CL/CLW.cl [hunk 6]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+        assign_##type(shared_mem + local_id, buffer[global_id]);\
+    }\
+    else\
+    {\
+        assign_##type(shared_mem + local_id, neutral_##bin_op##_##type);\
+    }\
+    \
+    if (global_id == 0)\
+    {\
+        *out = neutral_##bin_op##_##type;\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: What if global_id == 0 is executed last and you are doing atomic additions from other threads on non-initialized memory?

File: CLW/CL/CLW.cl [hunk 7]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+        atomic_##bin_op##_##type(out, shared_mem[0]);\
+    }\
+}
+
+// --------------------- NORMALIZATION ------------------------
+
+#define DEFINE_BUFFER_NORMALIZATION(type)\
+__kernel void buffer_normalization_##type(__global type* input,\
+                                          __global type* output,\
+                                          int buffer_count,\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: bad name

File: CLW/CL/CLW.cl [hunk 8]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+        min_buffer[local_id] = input[global_id];\
+        max_buffer[local_id] = input[global_id];\
+    }\
+    else\
+    {\
+        min_buffer[local_id] = neutral_min_##type;\
+        max_buffer[local_id] = neutral_max_##type;\
+    }\
+    \
+    if (global_id == 0)\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: will not work again, same reason

File: CLW/CL/CLW.cl [hunk 9]
```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+    }\
+    \
+    if (local_id == 0)\
+    {\
+        atomic_min_##type(auxiliary_buf, min_buffer[0]);\
+        atomic_max_##type(auxiliary_buf + 1, max_buffer[0]);\
+    }\
+    \
+    barrier(CLK_LOCAL_MEM_FENCE);\
+    type diff = auxiliary_buf[1] - auxiliary_buf[0];\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: again, global sync is required, not going to work

File: CLW/CL/CLW.cl (hunk 3 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Deleted

File: CLW/CL/CLW.cl (hunk 4 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Deleted

File: CLW/CL/CLW.cl (hunk 5 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Done

File: CLW/CL/CLW.cl (hunk 6 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Rewrited

File: CLW/CL/CLW.cl (hunk 7 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Rewrited

File: CLW/CL/CLW.cl (hunk 8 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Rewrited

File: CLW/CL/CLW.cl (hunk 9 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Rewrited

File: CLW/CL/CLW.cl [hunk 10]
```diff
@@ -1588,27 +1558,21 @@ __kernel void reduction_##bin_op##_##type(__global type* buffer,\
     int local_id = get_local_id(0);\
     int group_size = get_local_size(0);\
     \
-    if (global_id < buf_count)\
+    if (global_id < count)\
     {\
-        assign_##type(shared_mem + local_id, buffer[global_id]);\
+        *(shared_mem + local_id) = buffer[global_id];\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: Why not shared_mem[local_id]?

File: CLW/CLWParallelPrimitives.cpp [hunk 11]
```diff
@@ -921,65 +920,181 @@ void CLWParallelPrimitives::ReclaimTempBuffer(std::map<size_t, CLWBuffer<T> > co
     collection[buffer.GetElementCount()] = buffer;
 }
 
+CLWEvent CLWParallelPrimitives::Copy(unsigned int deviceIdx, CLWBuffer<cl_int> input, CLWBuffer<cl_int> output, int numElems)
```
Comment (yozhijk) [inline CLW/CLWParallelPrimitives.cpp]: Why not using clEnqueueCopyBuffer though CLWContext

File: CLW/CLWParallelPrimitives.cpp [hunk 12]
```diff
@@ -921,65 +920,181 @@ void CLWParallelPrimitives::ReclaimTempBuffer(std::map<size_t, CLWBuffer<T> > co
...
+
 template <class T>
-CLWEvent CLWParallelPrimitives::Normalize(const char* kernelName, unsigned int deviceIdx, CLWBuffer<T> input, CLWBuffer<T> output, int numElems, int groupSize, CLWBuffer<T> cache)
+T CLWParallelPrimitives::GetMaxNum()
+{
+    return std::numeric_limits<T>::max();
+}
+
+
+template <>
```
Comment (yozhijk) [inline CLW/CLWParallelPrimitives.cpp]: Why do we need this functions in CLWPrimitives?

File: CLW/CLWParallelPrimitives.cpp [hunk 13]
```diff
@@ -921,65 +920,181 @@ void CLWParallelPrimitives::ReclaimTempBuffer(std::map<size_t, CLWBuffer<T> > co
...
-    return context_.Launch1D(0, NUM_BLOCKS * WG_SIZE, WG_SIZE, copyKernel);
+    context_.WriteBuffer<T>(deviceIdx, cache, &min, 1);
+
+   Reduction(minReductionKernelName,
+             0,
+             input,
+             numElems,
+             cache).Wait();
+
+    context_.ReadBuffer<T>(deviceIdx, cache, &min, 1).Wait();
```
Comment (yozhijk) [inline CLW/CLWParallelPrimitives.cpp]: You should keep this data in GPUbuffer and do not read them back to CPU.

File: CLW/CL/CLW.cl (hunk 10 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Done

File: CLW/CLWParallelPrimitives.cpp (hunk 13 above)
Comment (AvKhokhlov) [reply CLW/CLWParallelPrimitives.cpp]: Done

File: CLW/CLWParallelPrimitives.cpp (hunk 12 above)
Comment (AvKhokhlov) [reply CLW/CLWParallelPrimitives.cpp]: Just helpers, cause cl_float3

File: CLW/CLWParallelPrimitives.cpp (hunk 11 above)
Comment (AvKhokhlov) [reply CLW/CLWParallelPrimitives.cpp]: Not mine

File: CLW/CL/CLW.cl (hunk 1 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Done

File: CLW/CLWParallelPrimitives.cpp [hunk 14]
```diff
@@ -953,3 +934,166 @@ CLWEvent CLWParallelPrimitives::Copy(unsigned int deviceIdx, CLWBuffer<cl_int> i
...
+
+    normalizeKernel.SetArg(argc++, input);
+    normalizeKernel.SetArg(argc++, output);
+    normalizeKernel.SetArg(argc++, numElems);
+    normalizeKernel.SetArg(argc++, cache);
+
+    return context_.Launch1D(deviceIdx, NUM_BLOCKS * WG_SIZE, WG_SIZE, normalizeKernel);
+}
+
+//CLWEvent CLWParallelPrimitives::Normalize(unsigned int deviceIdx, CLWBuffer<cl_int> input, CLWBuffer<cl_int> output, int numElems)
```
Comment (yozhijk) [inline CLW/CLWParallelPrimitives.cpp]: Dead code

File: CLW/CL/CLW.cl [hunk 15]
```diff
@@ -1492,3 +1507,100 @@ __kernel void segmented_distribute_part_sum_int_nocut(
         }
     }
 }
+
+// --------------------- ATOMIC OPERTIONS ------------------------
+
+#define DEFINE_ATOMIC(operation)\
+    __attribute__((always_inline)) void atomic_##operation##_float(volatile __global float* addr, float value)\
+    {\
+        union{\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: { from the new line

File: CLW/CL/CLW.cl [hunk 16]
```diff
@@ -1492,3 +1507,100 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+        barrier(CLK_LOCAL_MEM_FENCE);\
+    }\
+    if (local_id == 0)\
+        atomic_##bin_op##_##type(out + out_offset, shared_mem[0]);\
+}
+
+// --------------------- NORMALIZATION ------------------------
+
+#define DEFINE_BUFFER_NORMALIZATION(type)\
+__kernel void buffer_normalization_##type(__global type* input,\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: input should be const

File: CLW/CL/CLW.cl [hunk 17]
```diff
@@ -1492,3 +1507,100 @@ __kernel void segmented_distribute_part_sum_int_nocut(
...
+                                          __global type* output,\
+                                          int count,\
+                                          __global type* storage)\
```
Comment (yozhijk) [inline CLW/CL/CLW.cl]: storage should be const

File: CLW/CLWParallelPrimitives.cpp (hunk 14 above)
Comment (AvKhokhlov) [reply CLW/CLWParallelPrimitives.cpp]: Forgot to uncomment the section

File: CLW/CL/CLW.cl (hunk 15 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Done

File: CLW/CL/CLW.cl (hunk 16 above)
Comment (AvKhokhlov) [reply CLW/CL/CLW.cl]: Done

**Pull request merged**